### PR TITLE
*: support for limiting the number of client connections

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Limiting the number of connections is to control that there should not be too many connections on a single server, resulting in resource occupation. This feature is similar to [max_connections](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_max_connections) in MySQL, but it only controls the current server.

### What is changed and how it works?
Adds a check on the connection count when a connection is established.

### Check List 

Tests 

 - unit test

Code changes

- Has interface methods change

Release note

 - Support for limiting the number of client connections

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -477,6 +479,7 @@ var defaultConf = Config{
 	SplitRegionMaxNum:            1000,
 	RepairMode:                   false,
 	RepairTableList:              []string{},
+	MaxConnections:               151,
```
Comment (lonng) [inline config/config.go]: Why `151`?

Comment (zimulala) [reply config/config.go]: It's 151 in our original code and in MySQL.

Comment (wwar) [reply config/config.go]: The reason is historical. When the apache default was 150, and MySQL was 100, there were cases where web pages would contain too many connection errors. MySQL raised its limit to Apache's +1. It does not necessarily need to be followed literally.

**Comment by wwar** [comment]
This has an important behavior difference from MySQL, which is that in MySQL the `max_connections` cap only applies to regular users. There will always be an extra connection for super privileged users so that an administrator can log in and kill connections if there has been a thundering herd problem.

In MySQL 8.0 this was extended to provide a separate admin port, which is a better implementation: https://dev.mysql.com/worklog/task/?id=12138

File: config/config.go [hunk 2]
```diff
@@ -104,6 +104,8 @@ type Config struct {
 	// RepairMode indicates that the TiDB is in the repair mode for table meta.
 	RepairMode      bool     `toml:"repair-mode" json:"repair-mode"`
 	RepairTableList []string `toml:"repair-table-list" json:"repair-table-list"`
+	// MaxConnections is the maximum permitted number of simultaneous client connections.
+	MaxConnections uint32 `toml:"max-connections" json:"max-connections"`
```
Comment (jackysp) [inline config/config.go]: I think the config file in MySQL is like TiKV in TiDB for `SET` statement. So we may not need to add a config for it.

File: sessionctx/variable/sysvar.go [hunk 3]
```diff
@@ -136,7 +136,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal | ScopeSession, "ndb_index_stat_option", ""},
 	{ScopeGlobal | ScopeSession, OldPasswords, "0"},
 	{ScopeNone, "innodb_version", "5.6.25"},
-	{ScopeGlobal, MaxConnections, "151"},
+	{ScopeGlobal, MaxConnections, strconv.FormatInt(DefMaxConnections, 10)},
```
Comment (jackysp) [inline sessionctx/variable/sysvar.go]: Maybe we should add a upgradeVer function to upgrade it to a large value, e.g. 1000.

**Comment by jackysp** [comment]
> This has an important behavior difference from MySQL, which is that in MySQL the `max_connections` cap only applies to regular users. There will always be an extra connection for super privileged users so that an administrator can log in and kill connections if there has been a thundering herd problem.
> 
> In MySQL 8.0 this was extended to provide a separate admin port, which is a better implementation: https://dev.mysql.com/worklog/task/?id=12138

Yes, it maybe works like the extra port in MariaDB or Percona. We could discuss it in another issue.

File: sessionctx/variable/sysvar.go (hunk 3 above)
Comment (jackysp) [reply sessionctx/variable/sysvar.go]: PTAL @coocood

**Comment by coocood** [comment]
The connection limit applies to a single server, I think put it into the config file is better.

**Comment by jackysp** [comment]
> The connection limit applies to a single server, I think put it into the config file is better.

But it will make the global variable confused the users.

**Comment by jackysp** [comment]
> The connection limit applies to a single server, I think put it into the config file is better.

OK, it is better to use a different name, e.g. `server-max-connections`

**Comment by zimulala** [comment]
@coocood @jackysp PTAL. I update it as a config variable.

File: config/config.go [hunk 4]
```diff
@@ -477,6 +479,7 @@ var defaultConf = Config{
 	SplitRegionMaxNum:            1000,
 	RepairMode:                   false,
 	RepairTableList:              []string{},
+	MaxServerConnections:         1024,
```
[the remark this answers is not in the document]
Comment (zimulala) [reply config/config.go]: Done. PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**